src/main/java/bio/terra/service/dataset/flight/create/DatasetCreateFlight.java

```diff
@@ -108,6 +114,11 @@ public DatasetCreateFlight(FlightMap inputParameters, Object applicationContext)
       addStep(
           new CreateDatasetGetOrCreateContainerStep(
               resourceService, datasetRequest, azureContainerPdao));
+
+      // Turn on logging and monitoring for the storage account associated with the dataset
+      azureStorageMonitoringProvider
+          .configureSteps(datasetRequest.isEnableSecureMonitoring())
+          .forEach(s -> this.addStep(s.step(), s.retryRule()));
```

Comment:
This is a nice idea but it does result in a duplicate code everywhere it's used. Instead you could pass `Flight`, or if you didn't want to expose the entire object, a lambda for `addStep()`. E.g.
```
      azureStorageMonitoringProvider
          .configureSteps(datasetRequest.isEnableSecureMonitoring(), this::addStep);
```

```
  public void configureSteps(boolean isSecureMonitoringEnabled, BiConsumer<Step, RetryRule> addStep) {
   [...]
    // Deploy a Log Analytics Workspace if it doesn't exist already
    addStep.accept(new CreateLogAnalyticsWorkspaceStep(monitoringService), noRetry);
    // Create a diagnostic setting for the storage account if it doesn't exist already
    // This is what tells the storage account to send logs to the Log Analytics Workspace
    addStep.accept(new CreateDiagnosticSettingStep(monitoringService), noRetry);
[...]
```

Reply:
Haha, I actually had that exact code in my initial implementation but it seemed:
- a little clumsy
- harder to test
- to go against the intended use of stairway for flight construction

Reply:
The effect of this change results in the same stairway API calls being made, so I don't see how this goes against the intended use of stairway. I don't see how it's harder to test either, although I couldn't find any unit tests for  `AzureStorageMonitoringProvider` or `DatasetCreateFlight` so I'm not sure what test changes would be needed.

To me the original code is clumsier but [de gustibus...](https://en.wikipedia.org/wiki/De_gustibus_non_est_disputandum). To be clear, here are the befores and afters:

before
```
      azureStorageMonitoringProvider
          .configureSteps(datasetRequest.isEnableSecureMonitoring())
          .forEach(s -> this.addStep(s.step(), s.retryRule()));
```
After
```
      azureStorageMonitoringProvider
          .configureSteps(datasetRequest.isEnableSecureMonitoring(), this::addStep);
```

Before
```
  public List<StepDef> configureSteps(boolean isSecureMonitoringEnabled) {
    List<StepDef> steps = new ArrayList<>();
    RetryRuleNone noRetry = getRetryRuleNone();
    RetryRuleExponentialBackoff expBackoffRetry = getDefaultExponentialBackoffRetryRule();

    // Deploy a Log Analytics Workspace if it doesn't exist already
    steps.add(new StepDef(new CreateLogAnalyticsWorkspaceStep(monitoringService), noRetry));
    // Create a diagnostic setting for the storage account if it doesn't exist already
    // This is what tells the storage account to send logs to the Log Analytics Workspace
    steps.add(new StepDef(new CreateDiagnosticSettingStep(monitoringService), noRetry));

    // The following steps are only required for FedRAMP compliance
    // They are not turned on in all cases due to cost
    if (isSecureMonitoringEnabled) {
      // Create a rule to send the Log Analytics logs to a central storage account where the logs
      // will be retained for longer than the default 90 day minimum that the Log Analytics
      // Workspace stores
      steps.add(new StepDef(new CreateExportRuleStep(monitoringService), noRetry));
      // Deploy a Sentinel Workspace if it doesn't exist already
      steps.add(new StepDef(new CreateSentinelStep(monitoringService), noRetry));
      // Add any rules to Sentinel that will detect events of interest.  It takes a little bit for
      // Sentinel to be available so adding a retry rule here.
      steps.add(new StepDef(new CreateSentinelAlertRulesStep(monitoringService), expBackoffRetry));
      // Add a notification playbook rule to the Sentinel instance.  This ensures that a Slack
      // notification is sent when an alert is triggered.
      steps.add(new StepDef(new CreateSentinelNotificationRuleStep(monitoringService), noRetry));
    }
    return steps;
  }
```

after
```
  public void configureSteps(boolean isSecureMonitoringEnabled, BiConsumer<Step, RetryRule> addStep) {
    RetryRuleNone noRetry = getRetryRuleNone();
    RetryRuleExponentialBackoff expBackoffRetry = getDefaultExponentialBackoffRetryRule();

    // Deploy a Log Analytics Workspace if it doesn't exist already
    addStep.accept(new CreateLogAnalyticsWorkspaceStep(monitoringService), noRetry);
    // Create a diagnostic setting for the storage account if it doesn't exist already
    // This is what tells the storage account to send logs to the Log Analytics Workspace
    addStep.accept(new CreateDiagnosticSettingStep(monitoringService), noRetry);

    // The following steps are only required for FedRAMP compliance
    // They are not turned on in all cases due to cost
    if (isSecureMonitoringEnabled) {
      // Create a rule to send the Log Analytics logs to a central storage account where the logs
      // will be retained for longer than the default 90 day minimum that the Log Analytics
      // Workspace stores
      addStep.accept(new CreateExportRuleStep(monitoringService), noRetry);
      // Deploy a Sentinel Workspace if it doesn't exist already
      addStep.accept(new CreateSentinelStep(monitoringService), noRetry);
      // Add any rules to Sentinel that will detect events of interest.  It takes a little bit for
      // Sentinel to be available so adding a retry rule here.
      addStep.accept(new CreateSentinelAlertRulesStep(monitoringService), expBackoffRetry);
      // Add a notification playbook rule to the Sentinel instance.  This ensures that a Slack
      // notification is sent when an alert is triggered.
      addStep.accept(new CreateSentinelNotificationRuleStep(monitoringService), noRetry);
    }
  }
```

Reply:
I was thinking because add step is private so you have to pass in the context.  I mean, it's fine, but I get the sense the idea is to just call add step from within the flight class.  I also just added a unit test for the class.

Reply:
Not sure what you mean about add step being private. The caller creates a lambda (e.g., `this::addStep`) and the callee has an object of type `BiConsumer`. So flights aren't part of the callee API and aren't part of the tests for `configureSteps()`. I'll post a comment on the test code with an example of how the test can be written.

Reply:
Oh I get that, I just found that more clunky